opm/autodiff/BlackoilAquiferModel_impl.hpp

```diff
@@ -0,0 +1,138 @@
+namespace Opm {
+
+
+    template<typename TypeTag>
+    BlackoilAquiferModel<TypeTag>::
+    BlackoilAquiferModel(Simulator& ebosSimulator,
+                      const ModelParameters& param,
+                      const bool terminal_output)
+        : ebosSimulator_(ebosSimulator)
+        , param_(param)
+        , terminal_output_(terminal_output)
+    {
+        // const auto& gridView = ebosSimulator_.gridView();
+
+        // number_of_cells_ = gridView.size(/*codim=*/0);
+        // global_nc_ = gridView.comm().sum(number_of_cells_);
+        gravity_ = ebosSimulator_.problem().gravity()[2];
+        init(ebosSimulator_, aquifers_);
+    }
+
+
+    // called at the end of a time step
+    template<typename TypeTag>
+    void
+    BlackoilAquiferModel<TypeTag>:: timeStepSucceeded(const SimulatorTimerInterface& timer)
+    {
+        for (auto aquifer = aquifers_.begin(); aquifer != aquifers_.end(); ++aquifer)
+        {
+            aquifer->after_time_step(timer);
+        }
+    }
+
+    template<typename TypeTag>
+    void
+    BlackoilAquiferModel<TypeTag>::
+    assemble( const SimulatorTimerInterface& timer,
+              const int iterationIdx                )
+    {
+        // We need to update the reservoir pressures connected to the aquifer
+        updateConnectionIntensiveQuantities();
+
+        if (iterationIdx == 0) {
+            // We can do the Table check and coefficients update in this function
+            // For now, it does nothing!
+            prepareTimeStep(timer);
+        }
+
+        assembleAquiferEq(timer);
+    }
+
+
+    template<typename TypeTag>

```

Comment:
What is the function `updateConnectionIntensiveQuantities` doing? Could you please provide some explanation here? I am not familiar with `ElementContext` related.  Maybe @andlaus can provide some light here.

Reply:
@GitPaean Yes, it's supposed to get the updated solution at the same iteration, this is required so that the aquifer model can get the most recent solution field connected to the model.

As a very brief background, the model implements the inflow, Q as a function of the pressure field, P, which is implicit.

$$Q_{ai} = f(p_i(t+\delta t) - p_i(t))$$